### ISSUE TITLE
Zsh completions doesn't display list of available commands when pressing tab

### ISSUE THREAD

**Issue by Farid-NL** [rbenv/rbenv#1565]
## Expected behaviour

1. Pressing `<TAB>` should display a list of commands

    _Example with `gem`:_

    ![image](https://github.com/rbenv/rbenv/assets/34426099/8afc110c-a11c-4211-a23f-3d12dd4d177d)

2. When typing a command and there's only one possible option, it should complete the whole command

    _Example with `gem`:_

    ```shell
    gem ins<TAB>

    # completes the command
    gem install
    ```
## Actual behaviour

1. Pressing `<TAB>` **DOES NOT** display a list of commands

    _Example:_

    ```shell
    rbenv <TAB>

    # it shows three dots `...` for like 0.3s and then does nothing
    rbenv 
    ```

2. When typing a command and there's only one possible option, it **DOES** complete the whole command

    _Example:_

    ```shell
    rbenv  ins<TAB>

    # completes the command
    rbenv install
    ```

## Some extra info

|rbenv version|OS|Shell|Shell config|
|-|-|-|-|
|1.2.0-87-ge8b7a27|Linux x86_64|zsh|zsh4humans|

The only thing related to rbenv in my `.zshrc` is this:

```shell
eval "$(~/.rbenv/bin/rbenv init - zsh)"
```

**Comment by mislav** [comment]
Hi, thanks for reporting. It looks like your zsh shell has lots of customizations. Could it be that one of these customizations is affecting the behavior of typing `rbenv <TAB>`?

Here is what I get when I press Tab:
```
$ rbenv
--version           global              install             root                version             version-name        which
commands            help                local               shell               version-file        version-origin      
completions         hooks               prefix              shims               version-file-read   versions            
exec                init                rehash              uninstall           version-file-write  whence
```

My `~/.zshrc`:
```sh
autoload -U compinit
compinit -i

autoload colors
colors
```

**Comment by Farid-NL** [comment]
It does have a lots of customization, mainly because of zsh4humans (sort of a zsh framework). Until now I've never experienced issues with completions, until now.

My guess is that rbenv use the old completion system for zsh, [compctl](https://zsh.sourceforge.io/Doc/Release/Completion-Using-compctl.html#Completion-Using-compctl) and until now I have used _- rather the programs I use, handle completions with -_ the ['new' system](https://zsh.sourceforge.io/Doc/Release/Completion-System.html) (completions are managed by underscore files, like _rbenv and are in the fpath).

Also, in my zsh I don't declare autoload -U compinit anywhere, this is manged my zsh4humans as far as I know.

I guess this is mainly a problem with my zsh config (zsh4humans).

So I will reach with them ans see if I can fix this. Thanks.

---

> My ~/.zshrc:
> 
> autoload -U compinit
> compinit -i
> 
> autoload colors
> colors

I tried this with no avail.

**Comment by Farid-NL** [comment]
Correction, zsh4humans **does** have the autoload line, in particular this one: `autoload -Uz +X compinit || return`.

---

With the help of [this](https://www.reddit.com/r/zsh/comments/iwbaju/convert_to_the_newer_completion_system/), and editing the completion file `rbenv.zsh` accordingly:

```diff
@@ -2,11 +2,10 @@ if [[ ! -o interactive ]]; then
     return
 fi
 
-compctl -K _rbenv rbenv
+autoload -U compinit
+compdef _rbenv rbenv
 
 _rbenv() {
-  local words completions
-  read -cA words
 
   emulate -L zsh
 
@@ -16,5 +15,5 @@ _rbenv() {
     completions="$(rbenv completions ${words[2,-2]})"
   fi
 
-  reply=("${(ps:\n:)completions}")
+  compadd - "${(ps:\n:)completions}"
 }

``` 

I got it working

![image](https://github.com/rbenv/rbenv/assets/34426099/751fe827-d6bd-4737-93eb-6dcc27cc8e0d)

---

I don't know if this will broke someone else completions, but maybe you can test it and if everything works it could be a potential fix for other people with this problem.

**Comment by mislav** [comment]
It's true that we still use the outdated `compctl` system and that we could potentially migrate to the new completion system, especially to make use of features like autoloading and descriptions.

However, the point that I was making is that even with the `compctl`-based approach, rbenv zsh completions _should_ work, as evident by them working for me in zsh 5.9 with a very minimal zsh setup. In fact, I've just verified it again: I've started an instance of zsh with no startup files (no zprofile or zshrc), loaded our (legacy) completion script, and it just works. I do not know why it doesn't work for you.

**Comment by Farid-NL** [comment]
> I've just verified it again: I've started an instance of zsh with no startup files (no zprofile or zshrc), loaded our (legacy) completion script, and it just works. I do not know why it doesn't work for you.

You're right, I tried this as well and it didn't work for me for some reason

---

> Could you run zmodload to check if zsh/compctl is listed?

It is listed, it's the first one to appear.

**Comment by mislav** [comment]
Well, that is really weird.

You're welcome to submit a PR with your changes that transition to compsys! I was hoping to explore improving zsh completions anyway by adding descriptions to completed commands.

**Comment by Farid-NL** [comment]
Thanks for taking your tome to reply btw!

I will submit a PR with those little changes mentioned above that transition to compsys. However it doesn't have descriptions. 

I'm actually working in that version (with descriptions), when I finish it (or at least most of it) I will submit another PR.

**Comment by mislav** [comment]
Thanks a lot!

I also had a code spike locally for adding descriptions, so I have added it to your PR. Please take a look before you spend time implementing the same! Of course, simpler approaches than mine are aways welcome.

**Comment by Farid-NL** [comment]
Fixed in #1569
